075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,368 @@
+It means that, after the scaling up is triggered, the new pods will take time to be up, running and ready, so it's likely that the auto-rebalancing will happen on a future reconciliation.
+
+The scaling down runs synchronously within a single reconciliation which is blocked on waiting for the nodes to be removed.
+In the `KafkaReconcile.reconcile()` method, it is executed via the `scaleDown()` method which updates the involved `StrimziPodSet`(s) so that the corresponding controller will take care of removing the nodes while the current reconciliation waits for the end of the operation.
+
+The idea is about adding a new `autoRebalancing()` method in the `KafkaReconcile.reconcile()` flow, taking care of starting a rebalancing and checking the status of an ongoing one across reconciliations.
+The `KafkaReconcile.reconcile()` flow, taking into account methods involved in the scaling and rebalancing, would be something like:
+
+* ... 
+* `scaleDown()`
+* ...
+* `podSet()`
+* ...
+* `autoRebalancing()`
+* ...
+
+> In case of scaling down, the rebalancing has to happen before removing brokers and not after that as it looks like from the above flow. Despite that, some internal state will help to have the right actions order across reconciliations as explained in the next sections.
+
+#### Auto-rebalancing on scaling up
+
+The following flow is triggered when the user increases the `replicas` field within a `Kafka` or `KafkaNodePool` custom resource in order to scale up the cluster by adding more brokers.
+
+1. On the **1st** reconciliation (scaling up is requested): 
+     * the `podSet()` method triggers the scaling up, as described in the previous section, but the reconciliation is not blocked on waiting for the newly added pods to be ready.
+     * the `autoRebalancing()` method detects that a scaling up was requested because of `KafkaCluster.addedNodes()` returning a non empty list with the brokers' IDs to be added to the cluster. Because the `Kafka.status.cruiseControl.autoRebalance.state` doesn't exist, it's a new operation so the state is updated as `ScalingUp` and the `add-brokers` with the brokers' IDs. Of course, it doesn't trigger any rebalance because the scaling up is going on asynchronously.
+2. From the **2nd** to the **(N-1)th** reconciliation (scaling up is going on, checks when it's done):
+     * the `podSet()` does nothing because no new brokers were added this time.
+     * the `autoRebalancing()` evaluates the `Kafka.status.cruiseControl.autoRebalance.state: ScalingUp`, so a scaling up is running, and it has to check if it's finished or not. This can be achieved looking at the involved `StrimziPodSet`(s) and comparing `readyPods` and `pods` fields in the corresponding status. If they differ, the scaling up is still running. It should not take long time but it could go across multiple reconciliation, so this check repeats.
+3. On the **Nth** reconciliation (scaling up is done, rebalancing can start):
```

Comment:
Ok, I guess this is wrong because of the comment on line 169?

Reply:
Not sure what's wrong with this line, even because Jakub's comment on line 169 was referring to scale down (and he was right) not scale up. The point 3 here is about having the scale up done and the optimization proposal can be requested. The fact that CC doesn't have enough metrics collected will just take the `KafkaRebalance` in `PendingProposal` because of "not enough data in window" error but from a rebalancing FSM perspective we are already in the proposal/rebalancing process.